src/scene/morph-instance.js

```diff
@@ -0,0 +1,137 @@
+Object.assign(pc, function () {
+
+    /**
+     * @class
+     * @name pc.MorphInstance
+     * @classdesc An instance of pc.Morph. Contains weights to assign to every pc.MorphTarget, manages selection of active morph targets.
+     * @param {pc.Morph} morph - The pc.Morph to instance.
+     */
+    var MorphInstance = function (morph) {
+        this.morph = morph;
+
+        // weights
+        this._weights = [];
+        this._dirty = true;
+
+        // weights of active vertex buffers in format used by rendering
+        this._shaderMorphWeights = new Float32Array(pc.MorphInstance.RENDER_TARGET_COUNT);          // whole array
+        this._shaderMorphWeightsA = new Float32Array(this._shaderMorphWeights.buffer, 0, 4);        // first 4 elements
+        this._shaderMorphWeightsB = new Float32Array(this._shaderMorphWeights.buffer, 4 * 4, 4);    // second 4 elements
+
+        // temporary array of targets with non-zero weight
+        this._activeTargets = [];
+
+        // pre-allocate array of active vertex buffers used by rendering
+        this._activeVertexBuffers = new Array(pc.MorphInstance.RENDER_TARGET_COUNT);
+    };
+
+    Object.defineProperties(MorphInstance, {
+        // number of vertex buffer / weight slots
+        RENDER_TARGET_COUNT: { value: 8 }
+    });
+
+    Object.assign(MorphInstance.prototype, {
+
+        /**
+         * @function
+         * @name pc.MorphInstance#destroy
+         * @description Frees video memory allocated by this object.
+         */
+        destroy: function () {
+            if (this.morph) {
+                this.morph.destroy();
+                this.morph = null;
+            }
+        },
+
+        /**
+         * @function
+         * @name pc.MorphInstance#getWeight
+         * @description Gets current weight of the specified morph target.
+         * @param {number} index - An index of morph target.
+         * @returns {number} Weight.
+         */
+        getWeight: function (index) {
+            return this._weights[index];
+        },
+
+        /**
+         * @function
+         * @name pc.MorphInstance#setWeight
+         * @description Sets weight of the specified morph target.
+         * @param {number} index - An index of morph target.
+         * @param {number} weight - Weight.
+         */
+        setWeight: function (index, weight) {
+            this._weights[index] = weight;
+            this._dirty = true;
+        },
+
+        /**
+         * @function
+         * @name pc.MorphInstance#update
+         * @description Selects active morph targets and prepares morph for rendering. Called automatically by renderer.
+         */
+        update: function () {
+
+            var targets = this.morph._targets;
+
+            // collect active targets
+            this._activeTargets.length = 0;
```

Comment:
it should be more efficient to:
```
this._activeTargets.length = targets.length
...
this._activeTargets[i] =...
```

You could also reuse the previous update's { absWeight: ... } objects

Reply:
good suggestion, done, please do one last review on this change